lib/IDE/CompletionLookup.cpp

```diff
@@ -873,19 +873,67 @@ void CompletionLookup::addVarDeclRef(const VarDecl *VD,
     Builder.addFlair(CodeCompletionFlairBit::ExpressionSpecific);
 }
 
+/// Return whether \p param has a non-desiable default value for code
+/// completion.
+///
+/// 'ClangImporter::Implementation::inferDefaultArgument()' automatically adds
+/// default values for some parameters;
+///   * NS_OPTIONS enum type with the name '...Options'.
+///   * NSDictionary and labeled 'options', 'attributes', or 'userInfo'.
+/// But code-completion doesn't want to make them defaulted when it is the
+/// first parameter, the parameter doesn't have an argument label, and the
+/// imported function base name ends with those words.
+///
+///   E.g. -(void)setOptions:(NSArray *) opts;
+///
+/// Note that '-(void)performWithOptions:(NSArray *) opts;' doen't match the
+/// condition because the base name of the function in Swift is 'peform'.
+bool isNonDesiableImportedDefaultArg(const ParamDecl *param) {
+  auto kind = param->getDefaultArgumentKind();
+  if (kind != DefaultArgumentKind::EmptyArray &&
+      kind != DefaultArgumentKind::EmptyDictionary)
+    return false;
+
+  if (!param->getArgumentName().empty())
+    return false;
+
+  auto *func = dyn_cast<FuncDecl>(param->getDeclContext());
+  if (!func->hasClangNode())
+    return false;
+  if (func->getParameters()->front() != param)
+    return false;
+  if (func->getBaseName().isSpecial())
+    return false;
+
+  auto baseName = func->getBaseName().getIdentifier().str();
+  switch (kind) {
+  case DefaultArgumentKind::EmptyArray:
+    return (baseName.endswith("Options"));
+  case DefaultArgumentKind::EmptyDictionary:
+    return (baseName.endswith("Options") || baseName.endswith("Attributes") ||
+            baseName.endswith("UserInfo"));
```

Comment:
Would there be cases where we need to check case-insenstively?

Reply:
I don't think there're such cases practically. Technically:
```
-(void)attributes:(NSDictionary *)attrs;
-(void)set_attributes:(NSDictionary *)attrs;
```
But these namings are not common in ObjC.

Reply:
I don't think there are such cases practically. Technically these are:

```objc
-(void)attributes:(NSDictionary *)attrs;
-(void)add_attributes:(NSDictionary *)attrs;
```
But these namings are not common in ObjC.

Reply:
That sounds fair. I was thinking more of something like 
```objc
- (void)doSomething:(NSViewController *)viewController attributes:(NSDictionary *)attrs
```

But since we only care about the first argument, that doesn’t seem to be an issue.

Reply:
Worth noting that the implementation in ClangImporter actually assumes CamelCase, ie. it splits the words by camel case and then does case insensitive matching on them (see `ClangImporter::Implementation::inferDefaultArgument`). So Rintaro's first case would match, but eg. "userinfo" wouldn't (has to be "userInfo" or "UserInfo"). Would it be worth pulling that implementation out into a helper? Could maybe go in StringExtras?

Also I'm sure we spoke about this at some point, but the importer does this for *every* parameter. I assume considering the first argument here is just a heuristic that we think makes sense in most cases?